controllers/webhook.js

```diff
@@ -0,0 +1,260 @@
+            reqAuthenticated = false
+            if (!webhookSecretSent) {
+              // This could be worth logging... unless the endpoint gets hammered with spam.
+              errorsRaised.push('No secret found in the webhook request')
+            } else if (_verifyGitHubSignature(webhookSecretExpected, JSON.stringify(req.body), webhookSecretSent)) {
+              reqAuthenticated = true
+            } else {
+              // This could be worth logging... unless the endpoint gets hammered with spam.
+              errorsRaised.push('Unable to verify authenticity of request')
+            }
+          }
+
+          if (reqAuthenticated) {
+            await _handleMergeRequest(version, service, username, repository, branch, req.body).catch((errors) => {
+              errorsRaised = errors
+            })
+          }
+        }
+      }
+
+      break
+    case 'gitlab':
+      event = req.headers['x-gitlab-event']
+      if (!event) {
+        errorsRaised.push('No event found in the request')
+      } else {
+        if (event === 'Merge Request Hook') {
+          const webhookSecretExpected = config.get('gitlabWebhookSecret')
+          const webhookSecretSent = req.headers['x-gitlab-token']
+
+          let reqAuthenticated = true
+          if (webhookSecretExpected) {
+            reqAuthenticated = false
+            if (!webhookSecretSent) {
+              // This could be worth logging... unless the endpoint gets hammered with spam.
+              errorsRaised.push('No secret found in the webhook request')
+            } else if (webhookSecretExpected === webhookSecretSent) {
+              /*
+               * Whereas GitHub uses the webhook secret to sign the request body, GitLab does not.
+               * As such, just check that the received secret equals the expected value.
+               */
+              reqAuthenticated = true
+            } else {
+              // This could be worth logging... unless the endpoint gets hammered with spam.
+              errorsRaised.push('Unable to verify authenticity of request')
+            }
+          }
+
+          if (reqAuthenticated) {
+            await _handleMergeRequest(version, service, username, repository, branch, req.body).catch((errors) => {
+              errorsRaised = errors
+            })
+          }
+        }
+      }
+
+      break
+    default:
+      errorsRaised.push('Unexpected service specified.')
+  }
+
+  if (errorsRaised.length > 0) {
+    res.status(400).send({
+      errors: JSON.stringify(errorsRaised)
+    })
+  } else {
+    res.status(200).send({
+      success: true
+    })
+  }
+}
+
+const _verifyGitHubSignature = function (secret, data, signature) {
+  const signedData = 'sha1=' + crypto.createHmac('sha1', secret).update(data).digest('hex')
+  return bufferEq(Buffer.from(signature), Buffer.from(signedData))
+}
+
+const _handleMergeRequest = async function (version, service, username, repository, branch, data) {
```

Comment:
This is a pretty long function. We can probably break it down a bit.

Reply:
The removal of the comments shortened it a lot. I went further and extracted the GitHub-specific initialization out into a separate function.